aptos-move/aptos-vm/src/zkid_validation.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::move_vm_ext::{AptosMoveResolver, SessionExt};
+use anyhow::{Context, Result};
+use aptos_types::{
+    jwks::{jwk::JWK, PatchedJWKs},
+    on_chain_config::{CurrentTimeMicroseconds, OnChainConfig},
+    transaction::SignedTransaction,
+    vm_status::{StatusCode, VMStatus},
+    zkid::{ZkpOrOpenIdSig, MAX_ZK_ID_AUTHENTICATORS_ALLOWED},
+};
+use aptos_vm_logging::log_schema::AdapterLogSchema;
+use move_core_types::{language_storage::CORE_CODE_ADDRESS, move_resource::MoveStructType};
+
+fn get_current_time_onchain(resolver: &impl AptosMoveResolver) -> Result<CurrentTimeMicroseconds, VMStatus> {
+    CurrentTimeMicroseconds::fetch_config(resolver).ok_or_else(VMStatus::error(StatusCode::VALUE_DESERIALIZATION_ERROR, Some("could not fetch CurrentTimeMicroseconds on-chain config".to_string())))
+}
+
+fn get_jwks_onchain(resolver: &impl AptosMoveResolver) -> Result<PatchedJWKs> {
+    let bytes = resolver
+        .get_resource(&CORE_CODE_ADDRESS, &PatchedJWKs::struct_tag())?
+        .context("Failed to get jwks resource")?;
+    let jwks = bcs::from_bytes::<PatchedJWKs>(&bytes)?;
+    Ok(jwks)
+}
+
+pub fn validate_zkid_authenticators(
+    transaction: &SignedTransaction,
+    resolver: &impl AptosMoveResolver,
+    _session: &mut SessionExt,
+    _log_context: &AdapterLogSchema,
+) -> Result<(), VMStatus> {
+    // TODO(ZkIdGroth16Zkp): The ZKP/OpenID sig verification does not charge gas. So, we could have DoS attacks.
+    let zkid_authenticators =
+        aptos_types::zkid::get_zkid_authenticators(transaction).map_err(|_| {
+            VMStatus::error(
+                StatusCode::INVALID_SIGNATURE,
+                Some("Failed to fetch zkid authenticators".to_owned()),
+            )
+        })?;
+
+    if zkid_authenticators.len() > MAX_ZK_ID_AUTHENTICATORS_ALLOWED {
+        return Err(VMStatus::error(
+            StatusCode::TOO_MANY_ZK_AUTHENTICATORS,
+            None,
+        ));
+    } else if zkid_authenticators.is_empty() {
```

Comment:
Nit: separate this out as its own `if` conditional. And probably put it as the first conditional after instantiating `zkid_authenticators`